### ISSUE TITLE
sound is cut off with longer sound clips

### ISSUE THREAD

**Issue by danielb2** [ankidroid/Anki-Android#4457]
###### Research

_Enter an [ x ] character to confirm the points below:_

[ x ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[  x ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[ x ] I have searched for similar existing issues here and on the user forum

This has been an issue with a few versions now. I was lazy to file it thinking it would be obvious for other people. But I guess it's not so i've tried to be very detailed with reproduction.
###### Reproduction Steps
1. play file with longer sentence. I used [this deck](https://ankiweb.net/shared/info/114060567) with the phrase: 彼は真面目な学生です。(search for `He's a serious student`)
2. The last part of the sound is cut off for either reading or listening.
###### Expected Result

Be able to hear the whole sound clip just like the desktop version
###### Actual Result

the end of the sound clip is cut short.
###### Debug info

AnkiDroid Version = 2.7

Android Version = 6.0.1

ACRA UUID = dbe65079-5270-421e-9e36-e69b380fd3f1

**Comment by hssm** [comment]
I can't reproduce the problem with that deck.

**Comment by danielb2** [comment]
I'm using a OnePlus 3. Not sure if this makes a difference

**Comment by eginhard** [comment]
Fine on a OnePlus 1 as well. Have you noticed similar problems in any other apps?

**Comment by danielb2** [comment]
Other apps different from Anki? No, I have not, but the ones I have are Music and Google Play. Those are really the only one which do audio-playback.

I'll try a re-install

**Comment by danielb2** [comment]
same problem.

**Comment by danielb2** [comment]
[example.wav.zip](https://github.com/ankidroid/Anki-Android/files/546592/example.wav.zip)

She should say "gakusei desu"  but the "desu" is cut short.

**Comment by hssm** [comment]
The file you attached is cut short. So the problem isn't the playback of the file but the file itself.

How did you transfer the media to AnkiDroid?

**Comment by danielb2** [comment]
I don't know how else to put this, but of course it sounds like it's cut short. That's the whole subject of this issue. I recorded what I hear, and didn't record a bunch of empty sound afterwards. I can re-record and do that if that will help.

**Comment by danielb2** [comment]
I also mention in my comment that the "desu" is cut short

**Comment by danielb2** [comment]
To answer your question about transfer, it started sounding cut short after I did an update a two (I think) versions prior). I can roll back if you have an apkg for me to try. 

IOW, it was in an working state, I upgraded ankidroid, and then it started to sound cut short. I did delete the entire app to try it as I mentioned, but it didn't seem to re-download anything. And on all other devices (anki on macos) it sounds fine.

**Comment by hssm** [comment]
Sorry I had assumed that you copied the sound file from you media directory.

You can find previous releases [here](https://github.com/ankidroid/Anki-Android/tags). Released versions don't have a suffix.

**Comment by danielb2** [comment]
I just held the phone up to the microphone to illustrate the issue :)

OK, I will try it out and let you know. thanks.

**Comment by danielb2** [comment]
I've rolled back all the way to 2.4.3 so now I'm thinking the deck must have been corrupted somehow. But each time I uninstall the whole app and re-download, it's not requiring me to download any of the decks. How do I completely wipe all the data?

**Comment by danielb2** [comment]
I just deleted it with a file manger. redownloading everything

**Comment by hssm** [comment]
Try playing the sound file through your file manager outside of AnkiDroid and see what happens.

**Comment by danielb2** [comment]
Many of the mp3 are cut short actually there too. I get a message saying "Couldn't play the track you requested" after playing part of it.  How can I find the sound-file that belongs to the snippet I've been referring to?  I'm curious to get the md5 sum and compare to what it should be to see if the complete file has been downloaded and if it's just a playback issue or something else is going on.

**Comment by hssm** [comment]
If you open the note in the note editor you can see the filename in the field.

**Comment by danielb2** [comment]
Yeah, so the file is cut off exactly the same way from the FS too.

**Comment by hssm** [comment]
If you copy those files to your desktop, are they still cut off? I.e., is it the media player on the device that is the problem or the files themselves?

**Comment by danielb2** [comment]
If I copy to desktop it plays fine. It doesn't seem to matter which audio player I chose, always the same issue on the phone. So it looks like it's not an ankidroid issue so I'm going to close this issue. Thanks with the help.

Any ideas what could cause this?

**Comment by hssm** [comment]
I don't know what could be causing it. You're not the first person to report this type of problem. so it's not completely isolated.

**Comment by timor** [comment]
Confirmed.  Same behaviour for me

Android 6.0.1
AnkiDroid 2.8.4

Tried with stock Web View version and current Web View version, makes no difference.
Playing files e.g. with VLC on the device works.

**Comment by timor** [comment]
Argh! Sorry for resurrecting this.  Previous comment was meant to go to #4643.  Seems to be the same problem though.

**Comment by meandmymind** [comment]
Same here. Deck works fine in Anki Desktop version - no cuts. I deleted my AnkiDroid version of this deck, exported Desktop version and then imported it in AnkiDroid and get the same problem. I guess it's an import/export problem.

**Comment by mikehardy** [comment]
I think the last comment here, that it may be an export/import or sync problem is most likely. I'd be curious to see checksums on the problematic media files between ankidesktop and ankidroid - that would point precisely to the sync (or import) process corrupting the media files somehow

**Comment by mikehardy** [comment]
or maybe the m4a is actually corrupt slightly but in a way the other players can deal with? Could you check a hash on the file on the platforms to see if the same file propagated bit-for-bit to eliminate syncing error? a sha1 or md5 or something? if it's fine, then I would suspect either all m4a files have this error, or there will be an m4a utility somewhere that can inspect file structure and it will locate some issue with the file that the "ffmpeg clipping from m4a file" process has created. If you follow it through we may have an answer here...

**Comment by dqkk** [comment]
Okay, I did the test.  MD5 hashes match on both computers and the exported version from the iphone.

As a further experiment I reencoded the clip to another m4a file (ffmpeg w/ libfdk_aac).  This one plays on the iphone with no problem.  So it may be that there is some corruption issue with the problematic m4a files (original source is the current version of itunes on macos, which could be doing something funny; apple software used to be great, but lately not so much).  On the other hand the "problematic" file plays fine in every audio utility I have (vlc, itunes, quicktime player 7/10, transcribe, audacity, mplayer), and neither ffmpeg nor ffprobe flagged any problems when reading it.  So even if there is  some issue with the file(s), a bit more robustness in anki's audio library is arguably a good thing.

If it helps I can post the problematic and reencoded files; also if there is something that uses the same audio library as anki/ankidroid I could test the clip out there.

**Comment by mikehardy** [comment]
@dqkk you have earned a virtual beer, that is super thorough testing. Thank you! Posting the files would help as having the data handy for reproduction scenarios that are data-dependent is extremely valuable.

Unfortunately I'm looking at the code we use, and it's the stock MediaPlayer provided by the system, you can see it here:
https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/libanki/Sound.java#L303

Looks like there are a few options if you want to know concretely it is a platform issue, here is some code that is just a basic use of MediaPlayer to experiment:
https://github.com/nazmulidris/android-simple-mediaplayer

I'm also looking at what Google themselves are dealing with in their Exoplayer implementation and this is an area where we are exposed to handset manufacturer bugs. For example, I'd put a $0.01 that you have a Samsung phone, am I right or do I owe you?
https://github.com/google/ExoPlayer/issues/3386
https://github.com/google/ExoPlayer/issues/3449

This makes me think that applications that don't have issues have likely given up on platform decoding and bundle their own decoders (in the same way we bundle our own sqlite library to avoid issues with the bundled implementation)

If my hunch is right and this is a platform decoder issue, then a similar strategy (bundling a more tolerant decoder) would be the solution, though I'd have to look really carefully before we did that - depending on a 3rd party library at all is decision, even if the API it presented was fantastic and easy to integrate.

I'm interested in knowing exactly what it is though if you feel like trying to prove it's the system MediaPlayer

**Comment by dqkk** [comment]
Sorry, I seem to be causing confusion; to clarify, my issue is on the iOS Anki app, not Anki-Android, so everything I reported may be ultimately useless to you.  I posted the information here because this problem started recently for me, and this is only one of three threads I found where similar, recent problems were reports.  I was assuming the code bases are similar enough that the problem could be the same, but this may or may not be the case.  (If it's an issue with the system audio decoder then it's unlikely to be the same issue.  But it seems quite a coincidence that I started having this problem only recently after using Anki for years, and Anki-Android users started to report a similar problem at around the same time.)

I'm attaching the audio files; again, they're the same audio clip in m4a format.  The one that gets clipped was chopped out of an itunes file by ffmpeg; the one that works fine was reencoded by ffmpeg.

If someone using Anki-Android can see if these files behave in the same way on their phone, or can do the same checksum test on clips they're having trouble with, then this would be much more useful than the tests I did.  In the meantime I'll see if I can get access to an android phone to test to see if this is indeed a common issue.


[this-one-gets-clipped.m4a.zip](https://github.com/ankidroid/Anki-Android/files/2478187/this-one-gets-clipped.m4a.zip)
[this-one-works-fine.m4a.zip](https://github.com/ankidroid/Anki-Android/files/2478188/this-one-works-fine.m4a.zip)
